### PR TITLE
Port CUDA sources to MUSA in place

### PR DESCRIPTION
**Pull request by yeahdongcn**
## What

Port CUDA sources to MUSA **in place** as the single default path of the setuptools build:

- Drive `torch_musa`'s `SimplePorting` with `musa_dir_path == cuda_dir_path` and an identity extension map, so ported `.cu`/`.cuh` sources keep their original names and locations — no `<dir>_musa` mirror, no `.cu` to `.mu` / `.cuh` to `.muh` rename.
- Original `#include` and source paths therefore resolve as-is: no per-file include rewriting, and no second source tree that could cause cross-tree `#pragma once` ODR.
- Restrict in-place content substitution to portable C/C++/CUDA sources and headers, leaving build scripts, templates, docs, and MUSA-native `.mu`/`.muh` files byte-for-byte unchanged.
- Port each source's parent directory alongside the include roots.
- Move the `torch/headeronly/util/*` to `c10/util/*` header rewrites into the data-type mappings.
- Publish the fixed in-place behavior as `torchada 0.1.70`. No 0.1.70 artifact was published before this PR, so this remains the final release identity.

## Why

Consumers that drive this build (vLLM-MUSA, SGLang `sgl-kernel`) otherwise carry hand-maintained patches whose only job is to rewrite `#include` directives so mirrored and renamed ported sources resolve. Porting in place removes that class of downstream patch.

MUSA-native `.mu`/`.muh` files must not undergo CUDA-to-MUSA substitutions: they are already MUSA source, and rewriting their inline assembly caused the runtime regression found during on-device verification. A native MUSA source that includes a *ported third-party* header should use its real `.cuh` name; `.muh` remains reserved for MUSA-native headers.

## Verification

- Mapping suite: `83 passed`, 4 build-gated skipped.
- Version/platform suite: `10 passed`; changed-file Black and isort checks passed.
- Built `torchada-0.1.70-py3-none-any.whl` successfully and verified its wheel metadata reports version 0.1.70.
- Real `SimplePorting` container check: `.cu`/`.cuh`/`.h` ported in place; non-source and MUSA-native files remained byte-identical; no `<dir>_musa` mirror was created.
- Fresh vLLM-MUSA mp_31 builds succeeded locally and on an S5000 using this code.
- Single-GPU vLLM-MUSA smoke passed on one visible S5000 with TP=1, Qwen3-0.6B BF16, FlashAttention 3, and valid non-empty generation. The final 0.1.70 correction changes only version metadata/tests.

Cross-repository consumer: MooreThreads/vllm-musa#95.

### CODE REVIEW

File: src/torchada/_mappings/data_types.py [hunk 1]
```diff
@@ -30,4 +30,11 @@
     'CUDA_R_8F_E5M2': 'MUSA_R_8F_E5M2',
     'cuda_fp16.h': 'musa_fp16.h',
     'cuda_bf16.h': 'musa_bf16.h',
+    # torch_musa's torch::headeronly snapshot omits these scalar-type headers;
+    # the c10 variants carry the same definitions. Only the dtype headers move
+    # (Exception.h / ScalarType.h under torch/headeronly resolve on torch_musa).
+    '#include <torch/headeronly/util/Float8_e4m3fn.h>': '#include <c10/util/Float8_e4m3fn.h>',
+    '#include <torch/headeronly/util/Float8_e4m3fnuz.h>': '#include <c10/util/Float8_e4m3fnuz.h>',
+    '#include <torch/headeronly/util/BFloat16.h>': '#include <c10/util/BFloat16.h>',
+    '#include <torch/headeronly/util/Half.h>': '#include <c10/util/Half.h>',
```
Comment (yeahdongcn) [inline src/torchada/_mappings/data_types.py:39]: Why didn't we need these on the `vllm-musa` side previously?

File: src/torchada/utils/cpp_extension.py [hunk 2]
```diff
@@ -154,6 +176,59 @@ def open_with_surrogateescape(file, mode="r", *args, **kwargs):
...
+       other file is left byte-for-byte untouched (or copied verbatim if a
+       distinct mirror destination is still in use).
+
+    2. The stock method opens the destination with ``"w"`` (truncating) while
+       the source handle is open, zeroing the file when src == dst. Reading all
+       lines up front before opening the destination makes the in-place write
+       safe.
+    """
+
+    def modify_file(self, cuda_filepath, musa_filepath):
```
Comment (yeahdongcn) [inline src/torchada/utils/cpp_extension.py:323]: Is `modify_file` a complete rewrite?

File: src/torchada/utils/cpp_extension.py [hunk 3]
```diff
@@ -154,6 +176,59 @@ def open_with_surrogateescape(file, mode="r", *args, **kwargs):
...
+            if os.path.abspath(cuda_filepath) != os.path.abspath(musa_filepath):
+                import shutil
+
+                shutil.copyfile(cuda_filepath, musa_filepath)
+            return
+        with open(cuda_filepath, encoding="utf-8", errors="surrogateescape") as f:
+            lines = f.readlines()
+        out = []
+        for line in lines:
+            if line.startswith("*") or line.startswith("/") or line == "":
```
[the remark this answers is not in the document]
Comment (yeahdongcn) [reply src/torchada/utils/cpp_extension.py:348]: This is inherited verbatim from torch_musa's own `SimplePorting.modify_file` — I kept its per-line skip when rewriting the method for in-place safety + the extension allowlist, so porting stays byte-compatible with upstream here. It only skips a line whose **first** character is `*` or `/` (block-comment bodies, `//`/`/*` at column 0). Translation-unit statements are indented inside functions, so a column-0 `*ptr = …` doesn't occur in the csrc we port, and both the vllm-musa and sgl-kernel trees port + compile cleanly. Leaving it as-is to preserve parity with the upstream heuristic; stricter comment detection would be a separate change to torch_musa's porter.

File: src/torchada/utils/cpp_extension.py [hunk 4]
```diff
@@ -788,175 +875,108 @@ def build_extensions(self):
...
+                    # the read-all modify_file patch makes src == dst writes safe.
+                    sp.musa_dir_path = source_dir
+                    sp.run()
+
+                    # SimplePorting.__init__ eagerly created an empty <dir>_musa; drop it.
+                    import shutil
+
+                    leftover = source_dir + "_musa"
+                    if os.path.isdir(leftover):
+                        shutil.rmtree(leftover, ignore_errors=True)
```
[the remark this answers is not in the document]
Comment (yeahdongcn) [reply src/torchada/utils/cpp_extension.py]: Good catch — fixed in 11dd2d0. Switched `shutil.rmtree(..., ignore_errors=True)` → a guarded `os.rmdir()`, which refuses a non-empty directory, so this line can only remove the **empty** `<dir>_musa` that `SimplePorting.__init__` eagerly creates, and can never force-delete real content. (`SimplePorting.__init__` itself rmtree+recreates `<dir>_musa` before `run()` — upstream torch_musa behavior — but in the in-place flow `run()` writes into the source dir, so the `<dir>_musa` we clean up is always the empty one.)


_Re-verified on-device (leased S5000, native `mp_31`) with `torchada 0.1.70`: vllm-musa `csrc` + flashinfer build = **BUILD rc=0**, all extensions import (`vllm._C` / `_moe_C` / `_C_stable_libtorch` / `vllm_musa._C`), compiled from in-place paths (no `<dir>_musa` mirror)._

File: src/torchada/utils/cpp_extension.py [hunk 5]
```diff
@@ -788,175 +875,108 @@ def build_extensions(self):
...
+                    return source_dir
+
+
+                @staticmethod
+                def _dir_has_cuda_headers(path):
+                    """True if ``path`` recursively holds any C/C++/CUDA header or source."""
+                    try:
+                        for _root, _dirs, files in os.walk(path):
+                            for f in files:
+                                if f.endswith((".h", ".hpp", ".cuh", ".cu")):
```
[the remark this answers is not in the document]
Comment (yeahdongcn) [reply src/torchada/utils/cpp_extension.py]: Fixed in 11dd2d0 — renamed `_dir_has_cuda_headers` → `_dir_has_portable_sources`, and it now tests the full `_PORTABLE_SOURCE_EXTS` set (the same allowlist `modify_file` uses), so a directory of only `.cc`/`.cpp` sources is no longer skipped by the port gate.


_Re-verified on-device (leased S5000, native `mp_31`) with `torchada 0.1.70`: vllm-musa `csrc` + flashinfer build = **BUILD rc=0**, all extensions import (`vllm._C` / `_moe_C` / `_C_stable_libtorch` / `vllm_musa._C`), compiled from in-place paths (no `<dir>_musa` mirror)._

File: src/torchada/_mappings/data_types.py (hunk 1 above)
Comment (yeahdongcn) [reply src/torchada/_mappings/data_types.py:39]: We did need them — previously as vllm-musa csrc patches (the `<torch/headeronly/util/*>` → `<c10/util/*>` include rewrites in patches 0047/0048). Moving them into torchada's data-type mappings makes the rewrite happen automatically during porting, so every consumer (vllm-musa and sgl-kernel) gets it without a per-repo patch — which is exactly what lets vllm-musa#95 delete 0047/0048.

File: src/torchada/utils/cpp_extension.py (hunk 2 above)
Comment (yeahdongcn) [reply src/torchada/utils/cpp_extension.py:323]: It replaces the method body, but it's not a from-scratch reimplementation: it keeps torch_musa's original per-line substitution + `cub/` skip + comment-line skip, and adds two things — (1) read the whole source before opening the destination, so an in-place port (dst == src) can't truncate the file mid-read; (2) the extension allowlist, so only compiled C/C++/CUDA sources/headers are substituted and build scripts/templates/docs in a ported directory are left byte-for-byte unchanged.

**Comment by yeahdongcn** [comment]
**On-device verification caught a runtime regression — root-caused and fixed in `4a0126c` (torchada 0.1.70).**

The first in-place build passed build + import (rc=0) but produced degenerate garbage output (`"rounds rounds…"`) on-device for Qwen3-8B (bf16 **and** FP8), while the pre-PR mirror build served correctly on the same host/image/model.

**Root cause.** In-place porting walks every file under a ported include root (`csrc/` is one). Because `.mu`/`.muh` were in the porting allowlist, `SimplePorting` applied the CUDA→MUSA content mapping to the hand-written **MUSA-native** `csrc/musa/*.{mu,muh}` as well. The `_mappings/compiler.py` rule `'asm volatile' → 'if(0) asm volatile'` — which neutralizes un-assemblable **CUDA** PTX — then guarded `csrc/musa/vec_utils.muh`'s `vload16_byp_slc` inline-asm 16-byte load with `if(0)`, so the load never executes and `dst` is returned **uninitialized**. That garbage feeds RMSNorm (`fused_add_rmsnorm.mu`) / GEMV (`gemv.mu`) → garbage logits → repeated-token output. The mirror build escaped because it compiled the *untouched original* `.mu`/`.muh` (its corrupted copies landed in the unused `csrc_musa/` mirror).

Ruled out along the way: mcc codegen is byte-identical for `.cu` vs `.mu` under `-x musa`; the forward-path includes resolve to the single correct copy; a **ccache-disabled** rebuild still garbaged (so it was never stale objects).

**Fix (`4a0126c`).** Exclude `.mu`/`.muh` (MUSA-native — already MUSA, need no porting) from content substitution via a new `_MUSA_NATIVE_EXTS`, leaving them byte-identical. Genuinely-CUDA `.cu` under `csrc/musa/` (e.g. `quantization/*.cu` using `cudaStream_t`/`__nv_fp8_e4m3`) still port correctly. `90a43e7` aligns a stale mirror-era build test with the in-place contract.

**Verified on-device** (S5000, native mp_31, image `vllm-musa-0706-green:smoke`, torchada 0.1.70, 256-tok prompt, all else held constant): `PASS qwen3_8b_bf16 →Paris`, `PASS qwen3_8b_fp8 →Paris` (FLASH_ATTN), `PASS deepseek_v2_lite_fp8 →Paris` (FLASH_MLA), `PASS qwen35_35b_a3b_fp8 →Paris` (fused MoE); `vec_utils.muh` asm preserved (no `if(0)`).

**Comment by yeahdongcn** [comment]
Version correction applied in `906a2d7`: the final release remains `torchada 0.1.70` because no 0.1.70 artifact has been published yet.

- Runtime code is unchanged from the previously verified head.
- Version/platform tests: `10 passed`.
- Changed-file Black, isort, and `git diff --check` passed.
- Built `torchada-0.1.70-py3-none-any.whl` and verified its metadata reports version 0.1.70.
- vLLM-MUSA PR #95 now consistently requires `torchada>=0.1.70`.
